crates/globwatch/src/lib.rs

```diff
@@ -0,0 +1,361 @@
+                    }
+                }
+            }
+            WatcherChange::Exclude(glob, _) => {
+                for p in glob_to_paths(&glob) {
+                    // we don't care if this fails, it's just a best-effort
+                    watcher.unwatch(&p).ok();
+                }
+            }
+        }
+    }
+}
+
+fn get_flush_id(relative_path: &Path) -> Option<u64> {
+    relative_path
+        .file_name()
+        .and_then(|p| p.to_str())
+        .and_then(|p| p.parse().ok())
+}
+
+/// A configuration change to the watcher.
+#[derive(Debug)]
+pub enum WatcherCommand {
+    /// A change to the watcher configuration.
+    Watcher(WatcherChange),
+    /// A request to flush the watcher.
+    Flush(oneshot::Sender<()>),
+}
+
+/// A change to the watcher configuration.
+///
+/// This is used to communicate changes to the watcher
+/// from other threads. Can optionally contain the span
+/// that the change was made in, for tracing purposes.
+#[derive(Debug)]
+pub enum WatcherChange {
+    /// Register a glob to be included by the watcher.
+    Include(String, Option<Id>),
+    /// Register a glob to be excluded by the watcher.
+    Exclude(String, Option<Id>),
+}
+
+/// A sender for watcher configuration changes.
+#[derive(Debug, Clone)]
+pub struct GlobSender(tokio::sync::mpsc::Sender<WatcherCommand>);
+
+/// The server is no longer running.
+#[derive(Debug, Copy, Clone)]
+pub struct ConfigError;
+
+impl GlobSender {
+    /// Register a glob to be included by the watcher.
+    #[tracing::instrument(skip(self))]
+    pub async fn include(&self, glob: String) -> Result<(), ConfigError> {
+        trace!("including {:?}", glob);
+        self.0
+            .send(WatcherCommand::Watcher(WatcherChange::Include(
+                glob,
+                Span::current().id(),
+            )))
+            .await
+            .map_err(|_| ConfigError)
+    }
+
+    /// Register a glob to be excluded by the watcher.
+    #[tracing::instrument(skip(self))]
+    pub async fn exclude(&self, glob: String) -> Result<(), ConfigError> {
+        trace!("excluding {:?}", glob);
+        self.0
+            .send(WatcherCommand::Watcher(WatcherChange::Exclude(
+                glob,
+                Span::current().id(),
+            )))
+            .await
+            .map_err(|_| ConfigError)
+    }
+
+    /// Await a full filesystem flush from the watcher.
+    pub async fn flush(&self) -> Result<(), ConfigError> {
+        let (tx, rx) = oneshot::channel();
+        self.0
+            .send(WatcherCommand::Flush(tx))
+            .await
+            .map_err(|_| ConfigError)?;
+        rx.await.map_err(|_| ConfigError)
+    }
+}
+
+/// Gets the minimum set of paths that can be watched for a given glob,
+/// specified in minimatch glob syntax.
+///
+/// note: it is currently extremely conservative, handling only **, braces, and
+/// ?. any other case watches the entire directory.
```

Comment:
```suggestion
/// note: it is currently extremely conservative, handling only **, braces, and ?.
/// Any other case watches the entire directory.
```

`?` was lonely on its own line